### ISSUE TITLE
where(..., drop=True) error

### ISSUE THREAD

**Issue by pwolfram** [pydata/xarray#1350]
These results appear to be incorrect unless I'm missing something:
```python
In [1]: import xarray as xr

In [2]: import numpy as np

In [3]: array = xr.DataArray(np.zeros((1,2,3)), dims=['time','x','y'], coords={'x':np.arange(2)})

In [4]: array[0,1,1] = 1

In [5]: array.where(array !=0, drop=True)
Out[5]: 
<xarray.DataArray (time: 1, x: 1, y: 1)>
array([[[ 0.]]])
Coordinates:
  * x        (x) int64 1
Dimensions without coordinates: time, y

In [5]: array.where(array !=0, drop=True).values
Out[5]: array([[[ 0.]]])

In [7]: array.values[array.values !=0]
Out[7]: array([ 1.])
```

**Comment by shoyer** [comment]
Indeed, this definitely looks like a bug. My guess is that the logic for `drop=True` doesn't handle dimensions without coordinates properly.

**Comment by pwolfram** [comment]
As it turns out this was found this hunting a more complicated bug, but I couldn't easily reproduce it via DataArrays from scratch.  My guess is that solving this will *most likely* solve that issue:

```python
    ds = xr.open_dataset('mpaso.hist.0100-06-01_00000_potDensityOnly.nc')
    ds['latCell'] = ds.latCell*180.0/np.pi
    ds.set_coords('latCell', inplace=True)

    potDen = ds.potentialDensity.where(ds.potentialDensity != 0, drop=True)
```
yielding
```python
Traceback (most recent call last):
  File "./plot_potential_density.py", line 52, in <module>
    main()
  File "./plot_potential_density.py", line 16, in main
    potDen = ds.potentialDensity.where(ds.potentialDensity != 0, drop=True)
  File "/Users/pwolfram/anaconda/lib/python2.7/site-packages/xarray/core/common.py", line 627, in where
    outobj = self.sel(**indexers)
  File "/Users/pwolfram/anaconda/lib/python2.7/site-packages/xarray/core/dataarray.py", line 672, in sel
    result = self.isel(drop=drop, **pos_indexers)
  File "/Users/pwolfram/anaconda/lib/python2.7/site-packages/xarray/core/dataarray.py", line 657, in isel
    ds = self._to_temp_dataset().isel(drop=drop, **indexers)
  File "/Users/pwolfram/anaconda/lib/python2.7/site-packages/xarray/core/dataset.py", line 1119, in isel
    new_var = var.isel(**var_indexers)
  File "/Users/pwolfram/anaconda/lib/python2.7/site-packages/xarray/core/variable.py", line 548, in isel
    return self[tuple(key)]
  File "/Users/pwolfram/anaconda/lib/python2.7/site-packages/xarray/core/variable.py", line 378, in __getitem__
    values = self._indexable_data[key]
  File "/Users/pwolfram/anaconda/lib/python2.7/site-packages/xarray/core/indexing.py", line 423, in __getitem__
    return type(self)(self.array[key])
IndexError: shape mismatch: indexing arrays could not be broadcast together with shapes (1,) (235446,) (100,) 
```

It is possible that `ds.set_coords('latCell', inplace=True)` is part of the problem for this bug, so it may actually be a few problems, not just one.

**Comment by shoyer** [comment]
see https://github.com/pydata/xarray/pull/1361

**Comment by pwolfram** [comment]
Both cases are fixed by #1361.
